File: Project 4/Complex.java
public class Complex {
    private double real;
    private double imaginary;
    public Complex() {
        //a + bi => 0 + 0i
        this(0, 0);
    }

    public Complex(double r, double i) {
        //a + bi
        this.real = r;
        this.imaginary = i;
    }

    public Complex(double r) {
        //a + 0i
        this(r, 0);
    }

    //Complex x Complex
    public Complex add(Complex other) {
        //(a + bi) + (c + di) => (a + c) + (b + d)i
        return new Complex(real + other.real, imaginary + other.imaginary);
    }

    public Complex sub(Complex other) {
        //(a + bi) - (c + di) => (a + c) - (b + d)i
        return new Complex(real - other.real, imaginary - other.imaginary);
    }

    public Complex mul(Complex other) {
        //(a + bi) * (c + di) => (ac - bd) + (ad + bc)i
        return new Complex(real * other.real - imaginary * other.imaginary, real * other.imaginary + imaginary * other.real);
    }

    public Complex div(Complex other) {
        //(a + bi) / (c + di) => (ac - bd) + (bc + ad)i / (c^2 + d^2) 
        //=> (ac - bd)/(c^2 + d^2) + (bc - ad)i/(c^2 + d^2)
        return new Complex((real * other.real + imaginary * other.imaginary) / (other.real * other.real + other.imaginary * other.imaginary), (imaginary * other.real - real * other.imaginary) / (other.real * other.real + other.imaginary * other.imaginary));
    }

    //Complex x double <--> double x Complex
    //This language has no friend function so can only choose 1

    public Complex add(double other) {
        //(a + bi) + c => (a + c) + bi
        return new Complex(real + other, imaginary);
    }

    public Complex sub(double other) {
        ////(a + bi) - c => (a - c) + bi
        return new Complex(real - other, imaginary);
    }

    public Complex mul(double other) {
        //(a + bi) * c => (ac) + (bc)i
        return new Complex(real * other, imaginary * other);
    }

    public Complex div(double other) {
        //(a + bi) / c => a / c + (b / c)i
        return new Complex(real / other, imaginary / other);
    }



    public String toString() {
        return real + " + " + imaginary + "i";
    }

    public static void main(String[] args) {
        Complex a = new Complex(1, 2);
        Complex b = new Complex(3, 4);

        double d = 5;

        //Complex x Complex
        System.out.println(a + " + " + b + " = " + a.add(b));
        System.out.println(a + " - " + b + " = " + a.sub(b));
        System.out.println(a + " * " + b + " = " + a.mul(b));
        System.out.println(a + " / " + b + " = " + a.div(b));

        //Complex x double
        System.out.println(a + " + " + d + " = " + a.add(d));
        System.out.println(a + " - " + d + " = " + a.sub(d));
        System.out.println(a + " * " + d + " = " + a.mul(d));
        System.out.println(a + " / " + d + " = " + a.div(d));

        //double x Complex does not work because Java doesnt use operator overloading
    }
}

File: Project 4/complex.cc
#include <iostream>
#include <string>

class complex {
    private:
        double real;
        double imaginary;

    public:
        // a + bi => 0 + 0i
        complex() : real(0), imaginary(0) {}

        // a + bi
        complex(double r, double i) : real(r), imaginary(i) {}

        // a + 0i
        complex(double r) : real(r), imaginary(0) {}

        //Complex x Complex
        complex operator+(const complex& other) const {
            return complex(real + other.real, imaginary + other.imaginary);
        }

        complex operator-(const complex& other) const {
            return complex(real - other.real, imaginary - other.imaginary);
        }

        complex operator*(const complex& other) const {
            return complex(real * other.real - imaginary * other.imaginary, real * other.imaginary + imaginary * other.real);
        }

        complex operator/(const complex& other) const {
            return complex((real * other.real + imaginary * other.imaginary) / (other.real * other.real + other.imaginary * other.imaginary), (imaginary * other.real - real * other.imaginary) / (other.real * other.real + other.imaginary * other.imaginary));
        }

        //Complex x double
        complex operator+(double other) const {
            return complex(real + other, imaginary);
        }

        complex operator-(double other) const {
            return complex(real - other, imaginary);
        }

        complex operator*(double other) const {
            return complex(real * other, imaginary * other);
        }

        complex operator/(double other) const {
            return complex(real / other, imaginary / other);
        }

        //double x Complex
        friend complex operator+(double otherD, const complex& otherC) {
            return complex(otherD + otherC.real, otherC.imaginary);
        }

        friend complex operator-(double otherD, const complex& otherC) {
            return complex(otherD - otherC.real, otherC.imaginary);
        }

        friend complex operator*(double otherD, const complex& otherC) {
            return complex(otherD * otherC.real, otherD * otherC.imaginary);
        }

        friend complex operator/(double otherD, const complex& otherC) {
            return complex(otherD / otherC.real, otherD / otherC.imaginary);
        }

        friend std::ostream& operator<<(std::ostream& os, const complex& c) {
            os << c.real << " + " << c.imaginary << "i";
            return os;
        }
};

int main() {
    complex a(1, 2);
    complex b(3, 4);

    double d = 5;

    //Complex x Complex
    std::cout << a << " + " << b << " = " << a + b << std::endl;
    std::cout << a << " - " << b << " = " << a - b << std::endl;
    std::cout << a << " * " << b << " = " << a * b << std::endl;
    std::cout << a << " / " << b << " = " << a / b << std::endl;

    //Complex x double
    std::cout << a << " + " << d << " = " << a + d << std::endl;
    std::cout << a << " - " << d << " = " << a - d << std::endl;
    std::cout << a << " * " << d << " = " << a * d << std::endl;
    std::cout << a << " / " << d << " = " << a / d << std::endl;

    //double x Complex
    std::cout << d << " + " << a << " = " << d + a << std::endl;
    std::cout << d << " - " << a << " = " << d - a << std::endl;
    std::cout << d << " * " << a << " = " << d * a << std::endl;
    std::cout << d << " / " << a << " = " << d / a << std::endl;


    return 0;
}